Rename GPUExtent3DDict.depth to .depthOrArrayLayers

kainino0x opened the pull request:
Proposal.

Updates Extent3D as well, as it would be confusing if they didn't match.

Keeps a placeholder "depth" member with type undefined, to prevent accidental use of a silently-ignored dictionary entry.
Yes, this is weird, but WebIDL appears to allow it. If a browser implementation of WebIDL doesn't support it - Chromium's doesn't, unsurprisingly - it can be emulated by using another type (`int` or `any`, say) and then manually throwing a TypeError if the member is present.

See previous discussions:
https://github.com/gpuweb/gpuweb/issues/1372#issuecomment-768866675
https://github.com/gpuweb/gpuweb/pull/1081#issuecomment-696470308


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/1390.html" title="Last updated on Feb 4, 2021, 5:59 AM UTC (09ced26)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1390/d305f2c...kainino0x:09ced26.html" title="Last updated on Feb 4, 2021, 5:59 AM UTC (09ced26)">Diff</a>

Kangz commented:
I thought we discussed this before in the group and agreed to keep `depth`, but I can't find the minutes.

This proposal is a bit too much hand-holding for developers "remember that it is the depth or the array layer!" when instead it would be better to educate them that 2D arrays and 3D textures are mostly the same thing except for layout and the mip-levels in the `depth` dimension. The current unified "array depth" or "3D depth" is more elegant and causes less mental overhead for developers that know WebGPU (as in have used 2D arrays at least once).

Disallowing `depth` at the WebIDL is even more hand-holding an I don't think we want to have such "knowledge bandaids" in the API itself. At most it could be a console warning in browsers that want it, but it would be unfortunate to encode historical design process in the shape of the API.

kvark commented:
I do appreciate the rename to a more clear one. It's not a big issue for general ergonomics since most textures will not have the depth, and those who do - better know for sure what they opt into.

As for disallowing "depth" at WebIDL - that seems like and overkill. If somebody doesn't read the spec, the examples, and accidentally uses "depth", the worst they are going to face is one iteration where they discover that their view creation fails and fix it. There are no spooky actions in distance here, we don't need to worry about this.

kainino0x commented:
Updated to .depthOrArrayLayers and removed the .depth funny business.

I intend to do the .depth thing in the TypeScript bindings to achieve the same benefit for TypeScript users. That is the more reasonable place to put things like that.

kainino0x commented:
> the worst they are going to face is one iteration where they discover that their view creation fails and fix it. There are no spooky actions in distance here, we don't need to worry about this.

I forgot to reply about this, it would also change a multi-layer texture copy to copy only one layer (or slice). But I think it's okay. Browsers can even choose to implement warnings (by adding an `any depth;` field and checking it's not provided) if they want.

kainino0x commented:
meeting fyi: no objections :)
